streaming/src/main/scala/org/apache/spark/streaming/scheduler/StreamingListener.scala

```diff
@@ -77,6 +78,36 @@ trait StreamingListener {
   def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) { }
 }
 
+/**
+ * :: DeveloperApi ::
+ * A StreamingListener that estimates the number of elements that the previous
+ * batch would have processed, for each stream, if the duration of computation
+ * was one batchDuration.
+ * @param batchDuration The nominal (yardstick) duration for computation.
+ */
+@DeveloperApi
+class LatestSpeedListener(batchDuration: Duration) extends StreamingListener {
+  var latestTime : Option[Time] = None
+  var streamIdToElemsPerBatch: Option[Map[Int, Long]] = None
+
+  override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted){
+    this.synchronized{
```

Comment:
How do you establish the happens-before edge between readers and writers?

-- 
Cheers,
√